libraries/AP_HAL_ChibiOS/I2CDevice.cpp

```diff
@@ -0,0 +1,213 @@
+        businfo[i].dma_init();
+        /*
+          setup default I2C config. As each device is opened we will
+          drop the speed to be the minimum speed requested
+         */
+        businfo[i].i2ccfg.op_mode = OPMODE_I2C;
+        businfo[i].i2ccfg.clock_speed = 400000;
+        businfo[i].i2ccfg.duty_cycle = FAST_DUTY_CYCLE_2;
+    }
+}
+
+I2CDevice::I2CDevice(uint8_t busnum, uint8_t address, uint32_t bus_clock, bool use_smbus, uint32_t timeout_ms) :
+    _retries(2),
+    _address(address),
+    _use_smbus(use_smbus),
+    _timeout_ms(timeout_ms),
+    bus(I2CDeviceManager::businfo[busnum])
+{
+    set_device_bus(busnum);
+    set_device_address(address);
+    asprintf(&pname, "I2C:%u:%02x",
+             (unsigned)busnum, (unsigned)address);
+    if (bus_clock < bus.i2ccfg.clock_speed) {
+        bus.i2ccfg.clock_speed = bus_clock;
+        if (bus_clock <= 100000) {
+            bus.i2ccfg.duty_cycle = STD_DUTY_CYCLE;
+        }
+    }
+}
+
+I2CDevice::~I2CDevice()
+{
+    printf("I2C device bus %u address 0x%02x closed\n", 
+           (unsigned)bus.busnum, (unsigned)_address);
+    free(pname);
+}
+
+/*
+  allocate DMA channel
+ */
+void I2CBus::dma_allocate(void)
+{
+    i2cStart(I2CD[busnum].i2c, &i2ccfg);
+}
+
+/*
+  deallocate DMA channel
+ */
+void I2CBus::dma_deallocate(void)
+{
+    i2cStop(I2CD[busnum].i2c);
+}
+
+bool I2CDevice::transfer(const uint8_t *send, uint32_t send_len,
+                         uint8_t *recv, uint32_t recv_len)
+{
+    bus.dma_handle->lock();
+
+    if (_use_smbus) {
+        bus.i2ccfg.op_mode = OPMODE_SMBUS_HOST;
+    } else {
+        bus.i2ccfg.op_mode = OPMODE_I2C;
+    }
+    
+    if (_split_transfers) {
+        /*
+          splitting the transfer() into two pieces avoids a stop condition
+          with SCL low which is not supported on some devices (such as
+          LidarLite blue label)
+        */
+        if (send && send_len) {
+            if (!_transfer(send, send_len, nullptr, 0)) {
+                bus.dma_handle->unlock();
+                return false;
+            }
+        }
+        if (recv && recv_len) {
+            if (!_transfer(nullptr, 0, recv, recv_len)) {
+                bus.dma_handle->unlock();
+                return false;
+            }
+        }
+    } else {
+        // combined transfer
+        if (!_transfer(send, send_len, recv, recv_len)) {
+            bus.dma_handle->unlock();
+            return false;
+        }
+    }
+
+    bus.dma_handle->unlock();
+    return true;
+}
+
+bool I2CDevice::_transfer(const uint8_t *send, uint32_t send_len,
+                         uint8_t *recv, uint32_t recv_len)
+{
+    uint8_t *recv_buf = recv;
+    const uint8_t *send_buf = send;
+
+    bus.bouncebuffer_setup(send_buf, send_len, recv_buf, recv_len);
+
+    for(uint8_t i=0 ; i <= _retries; i++) {
+        int ret;
+        i2cAcquireBus(I2CD[bus.busnum].i2c);
+        // calculate a timeout as twice the expected transfer time, and set as min of 4ms
```

Comment:
min timeout of 4ms is just a default value?
What does it depend on?

Reply:
yes, its a default, its to account for response time of the device itself. The SMBus driver passes a higher value as they are slow.